layers/+distributions/spacemacs-base/funcs.el

```diff
@@ -1088,14 +1085,28 @@ Decision is based on `dotspacemacs-line-numbers'."
                (* 1000 (car (spacemacs/mplist-get dotspacemacs-line-numbers
                                                   :size-limit-kb)))))))
 
+;; mode in :enabled, not in :disabled ==> t
+;; mode not in :enabled, in :disabled ==> nil
+;; mode in :enabled, parent in :disabled ==> t
+;; parent in :enabled, mode in :disabled ==> nil
+;; not in :enabled, not in :disabled, :enabled is empty ==> t
+;; not in :enabled, not in :disabled, :enabled is not empty ==> nil
+;; both :enabled and :disabled are empty ==> t
 (defun spacemacs//linum-enabled-for-current-major-mode ()
   "Return non-nil if line number is enabled for current major-mode."
-  (let ((modes (spacemacs/mplist-get dotspacemacs-line-numbers
-                                     :enabled-for-modes)))
-    (memq major-mode modes)))
-
-(defun spacemacs//linum-disabled-for-current-major-mode ()
-  "Return non-nil if line number is disabled for current major-mode."
-  (let ((modes (spacemacs/mplist-get dotspacemacs-line-numbers
-                                     :disabled-for-modes)))
-    (memq major-mode modes)))
+  (let* ((enabled-for-modes (spacemacs/mplist-get dotspacemacs-line-numbers
+                                                  :enabled-for-modes))
+         (disabled-for-modes (spacemacs/mplist-get dotspacemacs-line-numbers
+                                                   :disabled-for-modes))
+         (enabled-for-parent (apply #'derived-mode-p enabled-for-modes))
+         (disabled-for-parent (apply #'derived-mode-p disabled-for-modes)))
+    (or
+     ;; current mode or a parent is in :enabled-for-modes, and there isn't a
+     ;; more specific parent (or the mode itself) in :disabled-for-modes
+     (and enabled-for-parent
+          (not (let ((major-mode enabled-for-parent))
+                 (derived-mode-p disabled-for-parent))))
+     ;; current mode (or parent) not explicitly disabled, and :enabled-for-modes
+     ;; not explicitly specified by user (meaning if it isn't explicitly
+     ;; disabled then it's enabled)
+     (and (null enabled-for-modes) (not disabled-for-parent)))))
```

Comment:
Well, it handles the case where a major-mode has a parent in `enabled-for-modes` and another parent in `disabled-for-modes`. One of those parents is a child of the other parent. If the enabled one derives from the disabled one, then `some-name` is non-nil and `spacemacs//linum-enabled-for-current-major-mode` should return `nil`.

I could write a helper function:
```elisp
(defun spacemacs/derived-mode-p (major-mode &rest modes)
  "Non-nil if the MAJOR-MODE is derived from one of MODES."
  ...)
```
And use it to avoid the awkward naming, hide the awkwardly implemented check, and improve readability (maybe). End result:
```elisp
(defun spacemacs//linum-enabled-for-current-major-mode ()
  "Return non-nil if line number is enabled for current major-mode."
  (let* ((enabled-for-modes (spacemacs/mplist-get dotspacemacs-line-numbers
                                                  :enabled-for-modes))
         (disabled-for-modes (spacemacs/mplist-get dotspacemacs-line-numbers
                                                   :disabled-for-modes))
         (enabled-for-parent (apply #'derived-mode-p enabled-for-modes))
         (disabled-for-parent (apply #'derived-mode-p disabled-for-modes)))
    (or
     ;; current mode or a parent is in :enabled-for-modes, and there isn't a
     ;; more specific parent (or the mode itself) in :disabled-for-modes
     (and enabled-for-parent
          (not (spacemacs/derived-mode-p enabled-for-parent disabled-for-parent))
     ;; current mode (or parent) not explicitly disabled, and :enabled-for-modes
     ;; not explicitly specified by user (meaning if it isn't explicitly
     ;; disabled then it's enabled)
     (and (null enabled-for-modes) (not disabled-for-parent)))))
```